### DISCUSSION TITLE
Sponsor AVRDUDE

### DISCUSSION THREAD

**Discussion by kristofmulier** [avrdudes/avrdude#1003] [General]
The work done for AVRDUDE is of utmost importance for millions of developers. Therefore, I think you should consider adding the GitHub sponsor option. I think it's a pretty new feature, it looks like this:

![image](https://user-images.githubusercontent.com/19362684/174434050-95cd39c5-ca02-4af7-b869-409711b75a18.png)

I'll be happy to make the first donation :-)

**Comment by dl8dtl** [comment]
Well, the question then is: what to do with the money?
Personally: it won't be able to buy me more time for the project.

**Comment by kristofmulier** [comment]
@dl8dtl ,
I know that donations for opensource projects typically don't make a lot of difference. Most people (and companies) just use opensource without giving any reward to the makers. However, it could be a way for those that care, to show their care.

**Comment by dl8dtl** [comment]
I do understand *that* part, but the question remains, what to do with the money …

**Comment by MCUdude** [comment]
> what to do with the money …

I'd argue that spending donated money to buy hardware to resolve reported issues is the way to go. For instance, there are lots of unresolved Xmega-related issues that can't really be fixed until we have the hardware to test with. In #435 @mcuee has done a great job and linked related issues.

**Comment by dl8dtl** [comment]
If anyone wants to tackle those issues, I guess it's easy enough to convince Microchip to donate a devboard.

IMHO, the project lacks man power rather than money.

**Comment by mcuee** [comment]
> If anyone wants to tackle those issues, I guess it's easy enough to convince Microchip to donate a devboard.

I will agree with that. Last time (around 2011) when Travis and I were working on libusb-win32 and libusbk projects,  For Microchip I had good connection at the time and I got them to donate MPLAB ICD3 and some other tools to Travis. I got PICkit 2 and mnay PIC chip samples from Microchip for my help on Microchip's pk2cmd tool. 

And for Atmel it was just some mailing list contact (Atmel was using libusb-win32). And both companies were very helpful. For example, Atmel provides AVR Dragon, AT90USBKey and UC3-A3 Xplained (the projects are USB centric).

Now Atmel is under Microchip and I think Microchip is pretty good in terms of supporting open source community.

> IMHO, the project lacks man power rather than money.

I will tend to agree as well. That is what we are facing in libusb, libusb-win32 and libusbk as well (I am the non-developer admin of these projects). For example, libusb has quite many issues and pull requests pending the involvement of developers/reviewers, especially on Windows side. 

We did collect two rounds of donation for libusb-win32/libusbk project previously in 2011 and 2014. That was to get the digital signature necessary for the kernel driver signing. Without the digital signature for libusb0.sys and libusbk.sys, you will not be able to use Zadig to install the driver necessary for libusb/avrdude/openocd/etc.

I think it is still okay to have some donation to help some developers to get the necessary tools.

**Comment by mcuee** [comment]
> > what to do with the money …
> 
> I'd argue that spending donated money to buy hardware to resolve reported issues is the way to go. For instance, there are lots of unresolved Xmega-related issues that can't really be fixed until we have the hardware to test with. In #435 @mcuee has done a great job and linked related issues.

I think it is still okay to have some donation to help some developers to get the necessary tools. Some tools are not provided by Microchip (eg: FTDI2232H/FTDI232H based tool) and some people may not want to get the help from Microchip.

My original intention was to use avrdude as a tool to help testing libusb/libftdi/hidapi (just like OpenOCD). But then I got more interested in avrdude as it seems the codes are easier to try to understand than libusb and OpenOCD. So I will at least try to understand the root cause of the issue and try to understand the patches (not to come out with patches).

I do not want to bother with Microchip this time so I am buying cheaper tools myself -- the more expensive will be two STK500 (programmer portion only) clones and one AVRISP MKII clone on order. I also bought the ATmega4809 Curiosity Nano board, ATmega328PB xPlained Mini board and the original Arduino Nano Every from Mouser. I am reluctant to buy the ATxmega related xPlained board (eg: A1U, A3BU and E5) as of now since they cost about US$55 each. So I choose to order a cheaper ATxmega32D4 breakout board. The other tools like Arduino clones and USBASP/USBispTiny are not that expensive to get.

**Comment by mcuee** [comment]
@dl8dtl 
Other than testing (not a developer so usually I can not come out with patches, but I can mess with Linux/macOS/Windows, and also FreeBSD when needed, and I have access to quite some tools), I can help to organize the github issues a bit better (to label them and link them, including closed issues) if you give me the access. I am doing this type of work for libusb, hidapi, pyusb and libusbdotnet.

For example, I think the issues number were not that manageable one month ago -- now it is much more managable at 114.

**Comment by dl8dtl** [comment]
> I can help to organize the github issues a bit better (to label them and link them, including closed issues) if you give me the access.

Offer happily accepted, thanks!

**Comment by dl8dtl** [comment]
> That was to get the digital signature necessary for the kernel driver signing.

That's certainly a good point, although I don't think needed for AVRDUDE right now.

**Comment by mcuee** [comment]
> > That was to get the digital signature necessary for the kernel driver signing.
> 
> That's certainly a good point, although I don't think needed for AVRDUDE right now.

Indeed. The direction for non-Microchip tools should be to go with WinUSB Driver, using WinUSB Compat ID so that Windows can automatically install the WinUSB driver. Microchip tools are mostly using WinUSB now.

Ref: @mariusgreuel has written it well here. And he has provided alternative FW for USBasp, FabISP/USBtinyISP and micronucleus.

https://github.com/avrdudes/avrdude/wiki/The-story-on-libusb-for-Windows#recommendation-to-usb-device-vendors
> If you require a generic USB driver (i.e. not a class driver, such as HID or CDC), choose WinUSB on Windows. If you choose to use the WinUSB driver, consider adding a [Microsoft OS descriptor](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-defined-usb-descriptors) to your device firmware, so that the WinUSB driver is loaded automatically. You should not require your users to use tools such as Zadig.

BTW, for USBtinyISP, Adafruit has a signed driver package for libusb0.sys and that should still work. They mentioned that their code signing certificate has expired and not planned to renew.
Ref: https://github.com/adafruit/Adafruit_Windows_Drivers/releases

Alternative FW with WinUSB Compat ID.
* https://github.com/dioannidis/usbasp (USB Composite Device with WinUSB and HID)
* https://github.com/mariusgreuel/USBasp (WinUSB)
https://github.com/mariusgreuel/FabISP (WinUSB)
https://github.com/mariusgreuel/micronucleus (WinUSB)

**Comment by mcuee** [comment]
> > I can help to organize the github issues a bit better (to label them and link them, including closed issues) if you give me the access.
> 
> Offer happily accepted, thanks!

Thanks. I start to label the current issues and those have just been closed.

**Comment by mcuee** [comment]
Maybe we can move this to Discussion.

**Comment by MCUdude** [comment]
@mcuee please do!

**Comment by mcuee** [comment]
From here: it is probably better to get official Microchip tools from Microchip. But certain tools may not be available from Microchip. So donation may be a way to get some of the more expensive tools.
* https://github.com/avrdudes/avrdude/discussions/1021#discussioncomment-3206414

For the xmega, Microchip seems to only offer `XMEGA A1U Xplained Pro` with ATxmega128A1U and `XMEGA-E5 XPLAINED` with ATxmega32E5. Probably they are not good enough for #337 and #404. I am not so sure if tey have some in-house boards with ATxmega256A3U or similar (for #337 and #404) and ATxmega128A4U or similar (for #289). Probably we have to buy from places like AliExpress.

Reference: ATxmega256A3U development board, roughly US$56 + Shipment + tax.
https://www.aliexpress.com/item/4001166867043.html

BTW, the same shop also has the cheaper ATxmega32A4 breakout board.
https://wildchip.aliexpress.com/store/group/ATXMEGA-mini-Board/5881453_517582356.html

**Comment by mcuee** [comment]
@dl8dtl 
> If anyone wants to tackle those issues, I guess it's easy enough to convince Microchip to donate a devboard.

Just wondering if you have good contacts within Microchip. It will be great to get some official tools from Microchip, for @MCUdude and I, and probably @stefanrueger as well if he wants.

Same as @MCUdude, I am not interested in money and I can spend small money on the tools as well. But it is good to have hardware tools for testing.

From @MCUdude 
https://github.com/avrdudes/avrdude/issues/1143
> I don't own an Xplained Pro board, so I can't test this.
> Ideally, I'd like to get an ATmega324PB Xplained Pro and an Xmega A1U Xplained Pro for testing, but I can't really justify spending the money.
> In https://github.com/avrdudes/avrdude/discussions/1003 we discussed how one can support the Avrdude project. Personally, I'm not interested in money, but hardware for testing is always helpful.

**Reply by mcuee** [reply]
@MCUdude and I have documented the tools and boards we have, I think two of us already cover majority of the HW supported by avrdude, but there are still some gaps.
* https://github.com/avrdudes/avrdude/discussions/1021#discussioncomment-3940622

**Reply by mcuee** [reply]
@MCUdude 
I contacted local Microchip and managed to borrow the PICkit 4 and ATtiny817 Xplained Pro from them. Unfortunately they do not have ATmega324PB Xplained Pro and the XmegaA1U Xplained Pro board.

**Comment by mcuee** [comment]
From:
* https://github.com/avrdudes/avrdude/pull/1507#issuecomment-1750437529

> Hi @mcuee
> 
> > I think we got the testing devices well covered among @MCUdude, @dl8dtl and me. You can refer to the following discussion. I have most of the lower cost tools (either from Taobao/AliExpress or from Mouser Singapore) but lack those higher end tools from Microchip.
> 
> Which higher-end tools from Microchip would you like to add to your collection? I'll see what I can afford financially and send it to you.
> 
> > BTW, I am located in Singapore (the little red dot in SouthEast Asia), so there will be international shipment cost involved.
> 
> That shouldn't be a problem. I have a friend in China who works in an electronics distribution company. Maybe he can send some material to you (and I pay the bill). Or I order it from Mouser Singapore and get it sent directly to you.
> 
> Kind regards, Kristof

Thanks. I do not need those higher end Microchip tools as @MCUdude and @dl8dtl cover them well. I do not plan to get the tools like "Microchip AVR JTAGICE MKII (obsolete), JTAG ICE 3 (obsolete), Atmel-ICE, Power Debugger". 

But if there are something to be desired from my side, it would be something like those from @MCUdude's Tindie Store. They are well-designed yet not readily available from other places.
https://www.tindie.com/stores/mcudude/

1) microUPDI programmer: US$14.00
2) DIP-40 Arduino compatible development board: US$45
3) SerialUPDI programmer: US$22.00
4) ATtiny Arduino compatible development board: US$25.00

**Reply by kristofmulier** [reply]
Hi @mcuee 

> [@mcuee] But if there are something to be desired from my side, it would be something like those from @MCUdude's Tindie Store. They are well-designed yet not readily available from other places.

I didn't know the Tindie Store is from @MCUdude. The store's name rang a bell - I've heard from it before. I love its products. Just out of curiosity: @MCUdude => you run this store? Cool!

> [@mcuee]
> 1. microUPDI programmer: US$14.00
> 2. DIP-40 Arduino compatible development board: US$45
> 3. SerialUPDI programmer: US$22.00
> 4. ATtiny Arduino compatible development board: US$25.00

Great. I'll be happy to sponsor the purchase of this hardware and the shipping costs too. It's my pleasure to help the AVRDUDE initiative - especially given the fact that we all benefit from your work (I'm cofounder of the free Embeetle IDE for microcontrollers, and we use AVRDUDE and OpenOCD to flash the hardware).

How should I proceed practically? Shall I just send the money to @MCUdude ? Or should I make an account on the Tindie Store and place an order? In that case, I'd need your address @mcuee.

Kind regards,
Kristof

**Reply by mcuee** [reply]
Let's hear from the preference from @MCUdude.

**Reply by MCUdude** [reply]
@kristofmulier yes, I run the store. Not to make money, but to tinker, design and make products that others finds useful and well though out. Feel free to purchase something if there are things in my store you really want or need, but I don't expect anyone to order something just to support my work.

> How should I proceed practically? Shall I just send the money to @MCUdude ? Or should I make an account on the Tindie Store and place an order?

Me personally, I don't need any money to continue contributing to this project. I'm doing this because I enjoy doing this on my spare time, as since I consider this a hobby, money doesn't matter. What _does_ matter is when I'll have to buy expensive hardware I don't need for anything else than helping out with testing, and that's where donations is nice.

We are several developers that add new features and helps out with testing, and I'm probably the one that has the most comprehensive hardware lineup. However, there are still things I wish I had but can't quite justify, since it will probably only be used fot testing Avrdude.

If you, or anyone else want to donate anything, here's a list of things I with I had:

* An Arduino Leonardo. I accidentally connected a 12V power supply to a 5V USB hub recently, and smoked _alot_ of boards. My Leonardo went up in smoke, and this would be very convenient to have now that [#1507](https://github.com/avrdudes/avrdude/pull/1507) needs to be properly tested across different hardware.
* An Arduino Micro (not Pro Micro). This uses a similar bootloader as the Leonardo, but with a different pin configuration. It would be great to verify that [#1507](https://github.com/avrdudes/avrdude/pull/1507) works with this board as well, even though I'm pretty sure it would.
* PICkit5 (not yet, but sometime in the future perhaps). The PICkit5 uses a different protocol than the PICkit4, and would require a completely new integration with Avrdude. Currently, the protocol the PK5 speaks are not available to the public, but eventually this will be available. I'm probably not the one that will create Avrdude support for the PK5 (IIRC is was one or more Atmel employees that added support for the current lineup (Atmel ICE, SNAP, Pickit4, etc.)), but when the day comes alot of testing would be needed. I already own a PICkit4 and a SNAP, so I can't really justify buying a PICkit5, becuase I don't really need one for my own projects.
* AVR-EB chips. These are brand new, and probably not able from others than Microchip direct. However, we recently added support for them ([#1491](https://github.com/avrdudes/avrdude/pull/1491)), but we didn't have any hardware to verify that they _actually_ works with Avrdude at the time.

Since I live in Norway, orders from Farnell (UK) takes less than 24 hours to arrive. That's pretty impressive!

**Reply by mcuee** [reply]
@kristofmulier

I think @MCUdude needs are more immediate needs and more important than my requests (which are just nice to have). Please go ahead and help him instead. Thanks.

**Reply by mcuee** [reply]
Same as Hans, I don't need any money to continue contributing to avrdude project and other Open Source projects I am involved in (eg: libusb, libusb-win32, libusbK, libusbdotnet, hidapi, pyusb, etc). I have been involved in Open Source projects since 2005. It has become my long term hobby as well. 

I have some freedoms to buy cheaper tools with the approval of my wife. :-) That is why I have many tools purchased from Taobao/AliExpress or Mouser Singapore (up to S$100 or about US$73 now), but not the few more expensive Microchip tools. Now she does not really complain that I buy stuffs, but rather she complains my lack of organization of the stuffs and lack of space at home. ;-) You know, in Singapore, we live in high rise buildings and the area of the flat is not big (100 sqm is about the average, new ones are even smaller).

**Reply by kristofmulier** [reply]
Hi @MCUdude and @mcuee ,

> [@mcuee] I think @MCUdude needs are more immediate needs and more important than my requests (which are just nice to have). Please go ahead and help him instead. Thanks.

Okay, will do 👍 

> [@MCUdude] If you, or anyone else want to donate anything, here's a list of things I with I had [...]

Great. I found the following products on Farnell:

- **Arduino Leonardo** -> Farnell ordercode: 2133071
- **Arduino Micro** -> Farnell ordercode: 2285194

And the following one on Microchip Direct:
- **PICkit5** -> https://www.microchip.com/en-us/development-tool/PG164150

I can't find the following product on Microchip Direct:
- **AVR-EB chips**

Can you share a  purchase link for the AVR-EB chips? Please also verify if the purchase links I found for the Arduino boards and the PICkit5 correspond to what you had in mind. You can then send me your address via mail:

kristof@embeetle.com

It's my pleasure to help the AVRDUDE community this way!

Kind regards,
Kristof

*PS: The contributions I can afford right now are rather small, because I pay this out of my own pocket. However, once our startup Embeetle IDE takes off we will certainly contribute more to the two projects that are very important to us - OpenOCD and AVRDUDE.*

**Reply by mcuee** [reply]
<del>AVR EA chip should be available from Microchip Direct. Example is the AVR16EA28-I/SP and AVR16EA28-I/SS.
https://www.microchip.com/en-us/product/AVR16EA28</del>

Edit: my mistake -- this is not for AVR EB chip.

**Reply by mcuee** [reply]
<del>There is also a demo board: AVR64EA48 CURIOSITY NANO EVALUATION KIT at US$24.02
https://www.microchip.com/en-us/development-tool/EV66E56A</del>

Edit: my mistake -- this is not for AVR EB chip.

**Reply by MCUdude** [reply]
@mcuee I already have an AVR-EA board. However, AVR-EB chips are available here:
https://www.microchipdirect.com/product/search/all/AVR16EB

**Reply by mcuee** [reply]
Sorry my mistake.

**Comment by mcuee** [comment]
@MCUdude 

The following offer is from @kristofmulier.

> Hi @stefanrueger , @MCUdude and @mcuee , Thank you very much for all your efforts to fix this issue. If any of you are not having the necessary boards to test, I'm willing to make an order at Farnell, Digikey or Mouser and have it sent over to your address.
> 
> Since I'm not familiar with the AVRDUDE source code, I figured that I can contribute something in this way.
> 
> Kind regards, Kristof

**Comment by mcuee** [comment]
I have updated my testing tools here. I got some tools from Microchip/Atmel last time for my involvement in other open source projects like libusb, libusb-win32 and pk2cmd. One old jtag2key is courtesy of Amontec many years ago for my involvement in OpenOCD. Then I got two tools on loan from Microchip Singapore in 2023.
* https://github.com/avrdudes/avrdude/discussions/1021#discussioncomment-3104217

@SpenceKonde is also very kind and sent me the following at the end of 2022.

Nice serialupdi adapter with pogo pins and pin headers
AVR64DD32 board
AVR32DD14 breakout board
ATTiny3227 breakout board
